server/handler.go

```diff
@@ -115,26 +153,32 @@ func (h *Handler) ComQuery(
 
 			r = nil
 			proccesedAtLeastOneBatch = true
-
 			continue
 		}
 
-		row, err := rows.Next()
-		if err != nil {
+		select {
+		case err = <-errchan:
 			if err == io.EOF {
-				break
+				break rowLoop
 			}
-
 			return err
-		}
+		case row := <-rowchan:
+			outputRow, err := rowToSQL(schema, row)
+			if err != nil {
+				close(quit)
+				return err
+			}
 
-		outputRow, err := rowToSQL(schema, row)
-		if err != nil {
-			return err
+			r.Rows = append(r.Rows, outputRow)
+			r.RowsAffected++
+		case <-timer.C:
+			if h.readTimeout != 0 {
```

Comment:
Doesn't _any_ timeout count as an error? Assuming we went on long enough to overrun the default 1h timer shouldn't that also be an error?

Reply:
No, 0 means that there is no timeout configured for the connection, so the socket will (hopefully) timeout with the default OS one, which is usually huge (ideally this should detect cases where the socket is dead like when you kill the client, but it doesn't, see #800 . But on second thoughts the wait, and thus the check to not wait in vain, could be slower so I'm leaving it at 1 minute.